### ISSUE TITLE
ADALClientId in Info.plist affects behavior even though aadClientIdOverride is set

### ISSUE THREAD

**Issue by shimizu-sec** [msintuneappsdk/ms-intune-app-sdk-ios#405]
**Describe the bug:**
I set IntuneMAMSettings.aadClientIdOverride, aadRedirectUriOverride and aadAuthorityUriOverride and no ADAL* in Info.plist, it works fine.
Strangely, if I just add ADALClientId in Info.plist, the behevior is changed.
When I set ADALClientId different from aadClientIdOverride, IntuneMAMEnrollmentManager.instance().loginAndEnrollAccount showed UI as expected but after I input userid and password, login procedure results in timeout.
If I added ADALRedirectUri (I meas set both ADALClientId and ADALRedirectUri in Info.plist), login procedure passed but after login acquireTokenSilent returns MSALErrorDomain + interactionRequired and acquireTokenInteractively resulted in 53009 error.

I know server behavior depends on server configuration. But the point is that just setting ADALClientId in Info.plist changes behavior even if IntuneMAMSettings.aadClientIdOverride is set.
I believe IntuneMAMSettings.aadClientIdOverride should override ADALClientId in Info.plist.

**To Reproduce**
1. Set IntuneMAMSettings.aadClientIdOverride and no ADALClientId in Info.plist, it works fine.
2. Just add ADALClientId in Info.plist with the Id other than the value set by aadClientIdOverride.
3. Login procedure in loginAndEnrollAccount results in timeout. 
4. Add ADALRedirectUri in Info.plist
5. loginAndEnrollAccount works but acquireTokenSilent returns MSALErrorDomain + interactionRequired and acquireTokenInteractively resulted in 53009 error, though if ADALClientId and ADALRedirectUri is not set acquireTokenSilent is succeeded.

**Expected behavior:**
Behavior should not be changed even if ADALClientId is set in Info.plist when IntuneMAMSettings.aadClientIdOverride is set.

**Screenshots and logs:**
No log but I can provide if requested

**Smartphone (please complete the following information):**
 - Device: iPad Pro 11 inch 3rd generation
 - OS: PadOS 17.3.1

**Intune App SDK for iOS (please complete the following information):**
I tried ms-intune-app-sdk-ios 18.1.2 first and behavior was not chaged with 19.1.0.

**Additional info:**
I'm debugging the behavior that in one environment acquireTokenSilent works but in another acquireTokenSilent returns MSALErrorDomain + interactionRequired and acquireTokenInteractively.
During this debug, I found this behevior.
I suppose the difference of server bahavior seems to be caused by the difference of server settings, but I belive the behavior above is not expected, or at least confusing.

**Comment by shimizu-sec** [comment]
Hi sbandarenko,

Thank you for your reply.
Could you let me know how to enable log output of your library?
I can provide caller log but I believe it is not exactly useful to you.

Thank you in advance.

**Comment by sbondarenko-msft** [comment]
> Hi sbandarenko,
> 
> Thank you for your reply. Could you let me know how to enable log output of your library? I can provide caller log but I believe it is not exactly useful to you.
> 
> Thank you in advance.

Hi,

The logging API is mentioned in our docs here:
https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-ios-appendix#how-can-i-troubleshoot-my-app

Thanks

**Comment by shimizu-sec** [comment]
Hi sbandarenko,

Thank you for your information.
I've sent log files directly to you.
Hope it helps.

Manabu

**Comment by ChismanRaheem** [comment]
HI @shimizu-sec looking at the logs provided we can see the AAD has responded with an error invalid_grant and a unsuccessful token response.  My understanding is that you are experiencing the issue when adding just ADALClientId in Info.plist.  Currently its recommended configuring the following setting with in the **IntuneMaMSetting**:

- ADALClientId
- ADALAuthority
- ADALRedirectUri
- ADALRedirectScheme
- ADALCacheKeychainGroupOverride

**ref**: https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-ios-phase2#configure-msal-settings-for-the-intune-app-sdk

**Comment by shimizu-sec** [comment]
Hi ChismanEaheem,

Thank you for your effort. I understand what you mean and I tried some more.

At first, I should clarify my motivation since it was not clear in my first post.
What I'd like to do is to switch ClientId and Authority dynaically. For this, I set default ClientId and Authority in Info.pilist, and if I need to override in source code, I set  IntuneMAMSettings.aadClientIdOverride and aadAuthorityUriOverride (I call IntuneMAMEEnrollmentManager.instance().deRegisterAndUnenrollAcount to switch).
But it does not work as expected.

What's happening is:
- If ClientId and Authority in Info.plist is different from aad*Override, loginAndEnrollAccount is succeeded but acquireTokenSilent returns MSALErrorDomain + interactionRequired and acquireTokenInteractively resulted in 53009 error.
- If ClientId and Authority in Info.plist is same as  aad*Override, loginAndEnrollAccount (this is not overriden), loginAndEnrollAccount and acquireTokenSilent are succeeded as expected.
- If no ClientId and Authority in Info.plist and only set  aad*Override, loginAndEnrollAccount (this is not overriden), loginAndEnrollAccount and acquireTokenSilent are succeeded as expected.

From this observation, I suppose settings in Info.plist is not overriden by aad*Override as expected.
I added ADALRedirectUri in Info.plist as you suggested, but the behavior is not changed.

I will send log output of the case above.
Or any suggention is welcome.

Thank you in advance.

**Comment by shimizu-sec** [comment]
I sent log output some days ago.
Do you have any update for that?

**Comment by shimizu-sec** [comment]
Hi ChismanRaheem,

Thank you for your investigation.
I understand what you mean. I'll check tenant configurations.

**Comment by shimizu-sec** [comment]
Hi ChismanRaheem,

I've confirmed the application is registered as a single tenant application.
Now there are 2 sites (I'm not sure this is the correct usage of the word but I mean there are 2 different contracts of MS365) to connect. It is the reason I need to switch dynamically. One is our customer's so it is default, and another is to test including the review by Apple.
I confirmed in both sites the application was registered as "single tenant" in "Supported account types" at Entra admin center.

I'd like to confirm one point. Do you mean the behavior seemed to be multitenant from your observation of my log output?
If so, though it is not expected, I'd like to try to implement to handle multitenant behavior.
I know if the setting is "multitenant", I should handle Authority URIs as you suggested. But I supposed singe tenant, I didn't implement as you suggested so far.

Regarding the timing setting AAD*overrides, I set them before enrollment.

Thank you in advance.

**Comment by shimizu-sec** [comment]
Hi,

I investigated your side log and found a suspicious behavior.
When I set ADALClientId as 99999999-9999-9999-9999-999999999999 in Info.plist and aadClientIdOverride as 11111111-1111-1111-1111-111111111111 before enrollment, I believe ClientId 9999... should not be used at all.
But in log, I can find 9999... in somewhere like:
```
[2024-02-28 14:40:03] (Default accessor) Looking for token with aliases (null), tenant (null), clientId 99999999-9999-9999-9999-999999999999, scopes (null), (null), error code: 0
[2024-02-28 14:40:05] (Default cache) Get app metadata entries with clientId 99999999-9999-9999-9999-999999999999, environment (null), (null), error code: 0
```
Because of this, it seems that credential cache cannot store the data which should be stored.
```
[2024-02-28 14:40:26] (MSIDAccountCredentialCache) performing filtering check on cached credential item with the following properties - client ID: 11111111-1111-1111-111111111111, target: (null), realm: (null), environment: login.windows.net, familyID: (null), homeAccountId: 12345678, enrollmentId: Masked(null), appKey: Masked(null), applicationIdentifier: Masked(null), tokenTyp    e: (null), (null), error code: 0
[2024-02-28 14:40:26] (MSIDCredentialCacheItem) cached item superset mismatch; cached item had both invalid clientID and familyID. actual client ID: 11111111-1111-1111-111111111111, expected client ID: 99999999-9999-9999-9999-999999999999, actual family ID: (null), expected family ID: 1, (null), error code: 0
[2024-02-28 14:40:26] (MSIDAccountCredentialCache) cached item had mismatching realm/clientId/familyId/target/requestedClaims with the credential query. excluding from the results., (null), error code: 0
```

If ADALClientId in Info.plist is not defined, this does not happen (because there is no information of 9999...).
```
[2024-02-28 14:35:59] (Default accessor) Looking for token with aliases (null), tenant (null), clientId 11111111-1111-1111-111111111111, scopes (null), (null), error code: 0
[2024-02-28 14:35:59] (Default cache) Get app metadata entries with clientId 11111111-1111-1111-111111111111, environment (null), (null), error code: 0
...
[2024-02-28 14:35:59] (MSIDAccountCredentialCache) performing filtering check on cached credential item with the following properties - client ID: 11111111-1111-1111-111111111111, target: (null), realm: (null), environment: login.windows.net, familyID: (null), homeAccountId: 12345678, enrollmentId: Masked(null), appKey: Masked(null), applicationIdentifier: Masked(null), tokenType: (null), (null), error code: 0
[2024-02-28 14:35:59] (MSIDCredentialCacheItem) cached item supserset match; actual client ID: 11111111-1111-1111-111111111111, expected client ID: 11111111-1111-1111-111111111111, actual family ID: (null), expected family ID: , (null), error code: 0
```

From these log, I suppose if ADALClientId in Info.plist is defined, "Default cache" and "Default accessor" are created and bound for ADALClientId. This causes problem since the data which is bound for overriden client id is not stored or is not handled correctly.
Can I set aadClientIdOverride before the cache is created, or can I force to recreate the cache after aadClientIdOverride is set? If not, I believe aadClientIdOverride does not work correctly if ADALClientId in Info.plist is defined.

**Comment by shimizu-sec** [comment]
I'm so connfused since there is no response after I submitted my comment 2weeks ago.
Because of this issue, my project is stucked over 2 month.
Could you let me know what is necessary of you to investigate more? I will provide any information you need.

**Comment by ChismanRaheem** [comment]
The problem you're encountering is most likely due to a known issue that has been addressed in the SDK update 19.2.0. You can access this update here: [msintuneappsdk/ms-intune-app-sdk-ios: Intune App SDK for iOS enables data protection and mobile app management features in iOS mobile apps with Microsoft Intune (github.com)](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios) on GitHub. The update ensures that the values for override authentication set dynamically will take precedence over the plist configurations. Upgrading to version 19.2.0 is expected to resolve the issue.

**Comment by shimizu-sec** [comment]
Thank you for your information.
We've confirmed SDK v19.20.0 works as we expected.
We can close this issue.
Thank you again.
